### PR TITLE
12_Martin_Vrachev_Homework_03

### PR DESCRIPTION
**Pull request by mvvrachev**

### CODE REVIEW

File: homework_03/12_Martin_Vrachev/circles.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        try:
+            assert isinstance(radius, float)
+        except:
+            print("The radius has to be float!")
+            exit()
+
+        self.center = center
+        self.radius = radius
+
+    def find_relative_position(self, circle):
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:49]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            assert isinstance(y, float)
+        except:
+            print("The point coordinates have to be float!")
+            exit()
+        
+        self.x = x
+        self.y = y
+
+class Circle:
+    def __init__(self, center, radius):
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:33]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        assert isinstance(a, Point)
+        assert isinstance(b, Point)
+    except:
+        print('The function takes points as parameters!')
+        exit()
+
+    return math.sqrt(pow(b.x - a.x, 2) + pow(b.y - a.y, 2))
+
+class Point:
+    def __init__(self, x, y):
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:21]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+from enum import Enum
+import math
+
+class RelativePosition(Enum):
+    NO_COMMON_POINTS = 1
+    TOUCHING = 2
+    INTERSECTING = 3
+    SAME = 4
+
+def distance(a, b):
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:10]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    def find_relative_position(self, circle):
+        try:
+            assert isinstance(circle, Circle)
+        except:
+            print('The function takes type Circle as a paramater!')
+            exit()
+          
+        # Same
+        if self.center.x == circle.center.x and self.center.y == circle.center.y:
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:57]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            return RelativePosition.SAME
+        
+        ab = abs(distance(self.center, circle.center))
+        radii_sum = self.radius + circle.radius
+        radii_sub = abs(self.radius - circle.radius)
+        
+        # No common points
+        if ab > radii_sum or ab < radii_sub:
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:65]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            return RelativePosition.NO_COMMON_POINTS
+
+        # Touching
+        if ab == radii_sum or ab == radii_sub:
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:69]: OK

File: homework_03/12_Martin_Vrachev/circles.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            return RelativePosition.TOUCHING
+
+        # Intersecting
+        if radii_sub < ab < radii_sum:
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/circles.py:73]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    assert pytest_wrapped_e.type == SystemExit
+
+def test_create_circle():
+    p = Point(5.5, 4.5)
+    c = Circle(p, 10.0)
+    assert isinstance(c, Circle)
+
+def test_create_invalid_circle():
+    with pytest.raises(SystemExit) as pytest_wrapped_e:
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:20]: It is not necessary to write this on every test. You could add it on the top of the class(which you don't have) as global variable.

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    c1 = Circle(Point(2.0, 1.0), 16.0)
+    c2 = Circle(Point(6.0, 4.0), 21.0)
+    assert c1.find_relative_position(c2) == RelativePosition.TOUCHING
+
+def test_intersecting():
+    c1 = Circle(Point(2.0, 1.0), 5.0)
+    c2 = Circle(Point(6.0, 4.0), 6.0)
+    assert c1.find_relative_position(c2) == RelativePosition.INTERSECTING
+
+def test_same():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:78]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+def test_touching():
+    c1 = Circle(Point(2.0, 1.0), 2.0)
+    c2 = Circle(Point(6.0, 4.0), 3.0)
+    assert c1.find_relative_position(c2) == RelativePosition.TOUCHING
+
+    c1 = Circle(Point(2.0, 1.0), 16.0)
+    c2 = Circle(Point(6.0, 4.0), 21.0)
+    assert c1.find_relative_position(c2) == RelativePosition.TOUCHING
+
+def test_intersecting():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:73]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+def test_no_common_points():
+    c1 = Circle(Point(2.0, 1.0), 1.0)
+    c2 = Circle(Point(6.0, 4.0), 2.0)
+    assert c1.find_relative_position(c2) == RelativePosition.NO_COMMON_POINTS
+
+    c1 = Circle(Point(2.0, 1.0), 9.0)
+    c2 = Circle(Point(6.0, 4.0), 20.0)
+    assert c1.find_relative_position(c2) == RelativePosition.NO_COMMON_POINTS
+
+def test_touching():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:64]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+def test_find_realtive_position_invalid_arguments():
+    with pytest.raises(SystemExit) as pytest_wrapped_e:
+        p = Point(5.6, 1.2)
+        c = Circle(p, 7.8)
+
+        c.find_relative_position('circle')
+    
+    assert pytest_wrapped_e.type == SystemExit
+
+def test_no_common_points():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:55]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    p2 = Point(-2.0, 2.0)
+    assert distance(p1, p2) == 5.0
+
+def test_invalid_distance_arguments():
+    with pytest.raises(SystemExit) as pytest_wrapped_e:
+        distance(2,3)
+    
+    assert pytest_wrapped_e.type == SystemExit
+
+def test_find_realtive_position_invalid_arguments():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:46]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+def test_distance():
+    p1 = Point(2.0, 1.0)
+    p2 = Point(6.0, 4.0)
+    assert distance(p1, p2) == 5.0
+
+    p1 = Point(2.0, -1.0)
+    p2 = Point(-2.0, 2.0)
+    assert distance(p1, p2) == 5.0
+
+def test_invalid_distance_arguments():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:40]: OK

File: homework_03/12_Martin_Vrachev/test_circles.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+    assert pytest_wrapped_e.type == SystemExit
+
+    with pytest.raises(SystemExit) as pytest_wrapped_e:
+        p = Point(1.2, 8.7)
+        c2 = Circle(p, 'radius')
+
+    assert pytest_wrapped_e.type == SystemExit
+
+def test_distance():
```
Comment (petardmnv) [inline homework_03/12_Martin_Vrachev/test_circles.py:31]: OK

**Review by petardmnv — APPROVED** [review]

File: homework_03/12_Martin_Vrachev/test_circles.py (hunk 9 above)
Comment (VayerMaking) [reply homework_03/12_Martin_Vrachev/test_circles.py:20]: agreed great tip

**Review by VayerMaking — COMMENTED** [review]
everything seems fine, some things could be removed for simplicity

**Pull request merged**